### PR TITLE
Add selection - Part 1

### PR DESCRIPTION
**Pull request by HKalbasi**
This PR adds:
* Selection with mouse
* Selection with arrow keys
* Select all
* Copy the text inside selection using osc52

To make selection with mouse work with scroll, I needed to make scroll not move the cursor, and to do that, I made a substantial change to the cursor. The cursor now stores its position in file instead of in screen, so you can scroll in a way that put cursor out of screen (you will get focus back to the cursor if you type something). With the changes made in the cursor and scroll, ox now behaves more similar to other editors.

This PR does not implement any feature related to editing the selection, for example deleting it or replacing it by typing. The reason is that it requires a massive amount of change to the undo/event system, which I think it doesn't worth it and I have different ideas about implementing undo. So I differed that part to a later PR.

My idea about undo is that the current event system which requires each event to become reversible will become increasingly more complex over time (currently the challenge is selection, but imagine multi cursor, automatic refactors from compilers and lsp, ...). Instead, we can take a snapshot from the editor state (file content, cursor position, selection, maybe offset, ...) frequently (for example at each event, or 1 second after idle, or something like that, we can look at what other editors do) and at undo we can easily switch between those snapshots. Thankfully, the `ropey` library supports O(1) clone for rope so we don't have much challenge for storing the snapshots efficiently. I think this makes undo less error prone and reduces its overhead on later features. If you are interested in that, I can do it in another PR.

### CODE REVIEW

File: config/.oxrc [hunk 1]
```diff
@@ -59,16 +71,13 @@ event_mapping = {
         editor:save_as()
     end,
     ["ctrl_a"] = function()
```
Comment (curlpipe) [inline config/.oxrc:73]: Perhaps save all should become Ctrl + Shift + A instead as Ctrl + A is now being used to select all

File: config/.oxrc [hunk 2]
```diff
@@ -59,16 +71,13 @@ event_mapping = {
...
     end,
     ["ctrl_q"] = function()
         editor:quit()
     end,
-    ["shift_left"] = function()
-        editor:previous_tab();
-    end,
-    ["shift_right"] = function()
-        editor:next_tab();
+    ["ctrl_c"] = function()
```
Comment (curlpipe) [inline config/.oxrc:79]: Should we add Ctrl + V for paste as well?

File: kaolinite/src/document.rs [hunk 3]
```diff
@@ -187,16 +189,20 @@ impl Document {
         }
     }
```
Comment (curlpipe) [inline kaolinite/src/document.rs:191]: Needs a docstring

**Comment by curlpipe** [comment]
This looks fantastic and works really well from my initial tests
I'm just going to make a few checks:

- What I'll need to update in the wiki and other documentation
- Make sure there aren't any bugs
- Add a few more handy keyboard shortcuts

Thank you so much for this contribution. The above tasks won't take too long, so this can be merged very soon!

**Comment by curlpipe** [comment]
Oh and as for your proposed changes to the undo system - it does sound appealing especially as nowadays we have tons of memory, so it's not a problem to store multiple copies of the same document.

There are just two things that make me a little more reserved towards using a snapshot approach:

1. If we want to implement macros (e.g. create groups of events that can be applied over and over again) then this can't be done through snapshots
2. If the user opens a file that is massive and edits it, it could end up having to store hundreds of copies of a file that is potentially multiple megabytes in size, which could add up very quickly.

A middle ground approach might be something like how git does diffs. It does it line by line. So maybe if a user edits a line, we only store the lines that were inserted / deleted. That would help with problem #2, but it still doesn't help with problem #1.

**Comment by HKalbasi** [comment]
> If the user opens a file that is massive and edits it, it could end up having to store hundreds of copies of a file that is potentially multiple megabytes in size, which could add up very quickly.

That's the magic of the persistant rope data structure. You can have multiple clone of a 1 Gigabyte rope, edit each of them independently, and it still takes 1 Gigabyte of memory. The trick is that clones will share their internal data with each other.

> If we want to implement macros (e.g. create groups of events that can be applied over and over again) then this can't be done through snapshots

Aren't events of a macro system inherently different with those of undo? For example we want to record a paste event as a paste in a macro but in undo we want to store the actual text and cursor position. I think it makes sense to go with snapshots for undo and keep a separate list of events when recording a macro.

**Comment by curlpipe** [comment]
> Aren't events of a macro system inherently different from those of undo?

Ah yeah, I hadn't considered that. I was thinking of adjusting the indices of the events (but that is quite messy on second thought).

**Comment by curlpipe** [comment]
I've done some works on this PR, fixed a few bugs and as far as I can tell, it works really really well.

I'm happy to merge now.

**Comment by HKalbasi** [comment]
Thanks for bug fixes and quick merge!

One small note: my implementation of copy used the osc52 ansi codes, but if I understand correctly, the current implementation uses a library which uses x11. If it is based on x11, it wouldn't work over ssh. Can you check if it works over ssh? (I don't have a system right now)

**Comment by curlpipe** [comment]
Good point!

I moved towards the library `copypasta_ext` as it supported paste and quite a few different platforms.

Judging from the [documentation](https://docs.rs/copypasta-ext/latest/copypasta_ext/index.html) of the library it includes a osc52 clipboard [here](https://docs.rs/copypasta-ext/latest/copypasta_ext/osc52/index.html)

From my understanding, this library will switch to osc52 if the x11, wayland, osx, windows clipboard etc etc aren't accessible - so I assume it will work over ssh?

**Comment by HKalbasi** [comment]
From the docs, it seems it would fallback to osc52, but from the source, I guess maybe not. A test would tell us.

**Comment by HKalbasi** [comment]
I tested it, and unfortunately it doesn't work over ssh. Copy shows an error and paste panics.

**Comment by curlpipe** [comment]
What is the error / panic message?

I suppose we could fallback to what you did originally if the copy pasta library can't handle it.

**Comment by HKalbasi** [comment]
> What is the error / panic message?

Panic at the paste was this:
```
panicked at src/editor.rs:534:24:
called `Option::unwrap()` on a `None` value
```
Which would be solved by avoiding the unwrap. The error message was the feedback from `OxError::Clipboard` so I assume something failed in copypasta.

> I suppose we could fallback to what you did originally if the copy pasta library can't handle it.

Isn't only the osc52 implementation enough? It is supported by many terminals (list of them [here](https://github.com/ojroques/vim-oscyank)). We can also do what zellij does, it supports osc52 by default and supports a "copy command" for people who can't switch to a terminal emulator with osc52 support.

I would like to avoid copypasta if possible, since it increases build time of ox significantly.

**Comment by HKalbasi** [comment]
By the way, I noticed that this PR broke some of the kaolinite tests.

**Comment by curlpipe** [comment]
Would it be possible to have paste through osc52?

As for the tests, yeah, they're probably going to have to be completely rewritten due to a lot of changes happening. That's something I'll probably get to at some point, I'll put it as high priority.

**Comment by HKalbasi** [comment]
> Would it be possible to have paste through osc52?

It *is* possible, but the support is not as universal, some terminals will open a permission dialog and/or restrict it to be after some user event, and it doesn't work in tmux and zellij (the X11 based paste doesn't work either, the problem is that when the trigger of the paste is from the app, tmux doesn't now which user requested the paste, so it will fail).

I think we can have a osc52 based `editor:paste` function, but don't bind it to `ctrl_v` in the default config, and instead encourage users to bind `ctrl_v` to paste in their terminal emulators. In that way, it will work even in tmux.

 > As for the tests, yeah, they're probably going to have to be completely rewritten due to a lot of changes happening. That's something I'll probably get to at some point, I'll put it as high priority.

Great! I will try to add a CI so breaking tests won't happen again in next PRs.

**Pull request merged**